File: dsa_in_java/sorting/MergeSort.java
/*
 * Topic: Merge Sort using Divide & Conquer
 * Language: Java
 *
 * Description:
 * Merge Sort is a Divide and Conquer algorithm that divides the array into halves,
 * recursively sorts each half, and then merges them back together in sorted order.
 *
 * Time Complexity:
 * Recursive Merge Sort → O(n log n)
 * Iterative Merge Sort → O(n log n)
 * (Both have the same asymptotic complexity, but recursive version is easier to implement.)
 *
 * Space Complexity:
 * O(n) – because of the temporary arrays used during merging.
 */

public class MergeSort {

    // Function to merge two sorted halves of the array
    public static void merge(int[] arr, int left, int mid, int right) {
        // Find sizes of two subarrays
        int n1 = mid - left + 1;
        int n2 = right - mid;

        // Create temp arrays
        int[] L = new int[n1];
        int[] R = new int[n2];

        // Copy data into temp arrays
        for (int i = 0; i < n1; i++)
            L[i] = arr[left + i];
        for (int j = 0; j < n2; j++)
            R[j] = arr[mid + 1 + j];

        // Merge the temp arrays
        int i = 0, j = 0;
        int k = left;

        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) {
                arr[k] = L[i];
                i++;
            } else {
                arr[k] = R[j];
                j++;
            }
            k++;
        }

        // Copy remaining elements of L[] if any
        while (i < n1) {
            arr[k] = L[i];
            i++;
            k++;
        }

        // Copy remaining elements of R[] if any
        while (j < n2) {
            arr[k] = R[j];
            j++;
            k++;
        }
    }

    // Recursive Merge Sort function
    public static void mergeSort(int[] arr, int left, int right) {
        if (left < right) {
            // Find the middle point
            int mid = left + (right - left) / 2;

            // Recursively sort the two halves
            mergeSort(arr, left, mid);
            mergeSort(arr, mid + 1, right);

            // Merge the sorted halves
            merge(arr, left, mid, right);
        }
    }

    // Function to print the array
    public static void printArray(int[] arr) {
        for (int num : arr)
            System.out.print(num + " ");
        System.out.println();
    }

    // Main function to test Merge Sort
    public static void main(String[] args) {
        int[] arr = { 38, 27, 43, 3, 9, 82, 10 };
        System.out.println("Original array:");
        printArray(arr);

        mergeSort(arr, 0, arr.length - 1);

        System.out.println("Sorted array:");
        printArray(arr);
    }
}

/*
Example Output:
------------------------
Original array:
38 27 43 3 9 82 10
Sorted array:
3 9 10 27 38 43 82
*/

File: dsa_in_js/array/sorting/merge_sort.js
/**
 * Merge Sort Algorithm (Divide and Conquer)
 * -----------------------------------------
 * Time Complexity:
 *   - Best: O(n log n)
 *   - Average: O(n log n)
 *   - Worst: O(n log n)
 *
 * Space Complexity: O(n)
 *
 * Author: Ankit Dand
 * Description: Implementation of Merge Sort in JavaScript
 */

/**
 * merge() function combines two sorted halves into a single sorted array.
 * @param {Array<number>} left - Left half (sorted)
 * @param {Array<number>} right - Right half (sorted)
 * @returns {Array<number>} Sorted merged array
 */
function merge(left, right) {
  const merged = [];
  let i = 0,
    j = 0;

  // Compare elements from both halves and push the smaller one
  while (i < left.length && j < right.length) {
    if (left[i] < right[j]) {
      merged.push(left[i]);
      i++;
    } else {
      merged.push(right[j]);
      j++;
    }
  }

  // Add remaining elements from left (if any)
  while (i < left.length) {
    merged.push(left[i]);
    i++;
  }

  // Add remaining elements from right (if any)
  while (j < right.length) {
    merged.push(right[j]);
    j++;
  }

  return merged;
}

/**
 * mergeSort() function recursively divides the array and merges sorted halves.
 * @param {Array<number>} arr - Array to be sorted
 * @returns {Array<number>} Sorted array
 */
function mergeSort(arr) {
  if (arr.length <= 1) return arr; // Base case: already sorted

  const mid = Math.floor(arr.length / 2);
  const left = arr.slice(0, mid);
  const right = arr.slice(mid);

  // Recursive sort and merge
  return merge(mergeSort(left), mergeSort(right));
}

// Example Demonstrations

const example1 = [38, 27, 43, 3, 9, 82, 10];
const example2 = [5, 4, 3, 2, 1];
const example3 = [12, 11, 13, 5, 6, 7];

console.log("🧩 Merge Sort Example 1:");
console.log("Input:", example1);
console.log("Sorted:", mergeSort(example1));

console.log("\n🧩 Merge Sort Example 2:");
console.log("Input:", example2);
console.log("Sorted:", mergeSort(example2));

console.log("\n🧩 Merge Sort Example 3:");
console.log("Input:", example3);
console.log("Sorted:", mergeSort(example3));
